modules/11_subclustering/subclustering.qmd

```diff
@@ -0,0 +1,676 @@
+- Add explanation which cluster is split into what
+
+```{r}
+#| label: subclustering_split_cluster_2
+#| eval: false
+
+# Simple subclustering: Use the existing neighbors graph and split a cluster into subclusters. This method is the fastest. However, note that the neighbors graph is based on the dimensionality reduction of the top variable features of the WHOLE dataset. It might not be suitable for more fine-grained substructures within clusters. In such cases, a full subclustering of the cluster should be performed. 
+
+# USER_INPUT:
+# - Set the name of the cluster to be split
+# - Set the name of the neighbors graph that was used for clustering
+#  (see sc@commands$FindClusters, typically "RNA_snn" or "RNA_pca_snn")
+# - Set the subclustering resolution
+# - Provide a list of markers to verify the results 
+cluster_to_split = "10"
+graph_name = "RNA_snn"
+clustering_resolution = 1
+known_markers = list(CelltypeA=c("F3", "SYT6"), 
+                     CelltypeB=c("WDR3", "REG4"))
+
+# Do subclustering
+set.seed(getOption("random_seed"))
+SeuratObject::Idents(sc) = "subclustering"
+sc = Seurat::FindSubCluster(sc, 
+                            cluster=cluster_to_split,
+                            graph.name=graph_name,
+                            subcluster.name="sub.cluster",
+                            resolution=clustering_resolution,
+                            algorithm=4)
+
+# Get names of subclusters
+subcluster_names = unique(sc$sub.cluster) %>% 
+  grep(paste0(cluster_to_split, "_\\d+$"), v=T, x=.) 
+index = gsub(".+_(\\d+)$", "\\1", subcluster_names) %>% as.integer()
+subcluster_names = subcluster_names[order(index)]
+
+# Decide which subcluster is what
+SeuratObject::Idents(sc) = "sub.cluster"
+
+# DotPlot of known markers
+DotPlot(sc, 
+        features=known_markers,
+        assay=default_assay,
+        idents=subcluster_names) + 
+  AddPlotStyle(title="Known markers per subcluster")
+
+# Also calculate top10 markers between subclusters as additional information
+if (length(subcluster_names) > 1) {
+  markers = purrr::map_dfr(subcluster_names, function(cl) {
+    num_cells = sum(SeuratObject::Idents(sc) == cl)
+    if (num_cells<3) return(NULL)
+    
+    other_cl = setdiff(subcluster_names, cl)
+    num_cells = sum(SeuratObject::Idents(sc) %in% other_cl)
+    
+    mrk = Seurat::FindMarkers(sc,
+                              assay=default_assay,
+                              ident.1=cl, 
+                              ident.2=other_cl, 
+                              only.pos=TRUE,
+                              )
+    mrk$cluster = cl
+    mrk$gene = rownames(mrk)
+    rownames(mrk) = NULL
+    mrk = head(mrk, 10)
+    return(mrk)
+  })
+  markers = markers %>% 
+    dplyr::group_by(cluster) %>%
+    dplyr::relocate(cluster,gene)
+  gt::gt(markers, caption="Top10 markers computed per subcluster") %>% 
+    tab_options(container.height=450)
+}
+
+# Decide which subcluster is what or repeat with different resolution
```

Comment:
What's this? Was this not done automatically previously?

Reply:
(Just trying to keep to a minimum what we have to adapt for every project and every cluster, and here we need to modify the 10 - and we have to remember that)

Reply:
Well this module cannot be fully automated - you have to run it at least once manually :-) .

My original intention was to name subclusters already at this step so that I can use them in another round of subclustering. So for example you would first split immune cells into Bcells and Tcells. Then you round subclustering on Tcells. This hierarchical approach worked quite well in the past.

But I have removed these lines of code now. At this point in the module, clusters are names 10_1, 10_2, ... . This means that we have to keep track of what is what.

At the end of the module, you will then have two possibilities:

a) Renumber all clusters automatically by decreasing size
b) Name clusters manually (e.g. cell type)